packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -14,54 +14,41 @@
  * limitations under the License.
  */
 
-import { Attributes, diag } from '@opentelemetry/api';
+import { Attributes, AttributeValue, diag } from '@opentelemetry/api';
+import { SDK_INFO } from '@opentelemetry/core';
 import {
-  SEMRESATTRS_SERVICE_NAME,
-  SEMRESATTRS_TELEMETRY_SDK_LANGUAGE,
-  SEMRESATTRS_TELEMETRY_SDK_NAME,
-  SEMRESATTRS_TELEMETRY_SDK_VERSION,
+  ATTR_SERVICE_NAME,
+  ATTR_TELEMETRY_SDK_LANGUAGE,
+  ATTR_TELEMETRY_SDK_NAME,
+  ATTR_TELEMETRY_SDK_VERSION,
 } from '@opentelemetry/semantic-conventions';
-import { SDK_INFO } from '@opentelemetry/core';
-import { defaultServiceName } from './platform';
 import { IResource } from './IResource';
+import { defaultServiceName } from './platform';
+import {
+  DetectedResource,
+  DetectedResourceAttributes,
+  MaybePromise,
+} from './types';
+import { isPromiseLike } from './utils';
 
-/**
- * A Resource describes the entity for which a signals (metrics or trace) are
- * collected.
- */
 export class Resource implements IResource {
-  static readonly EMPTY = new Resource({});
-  private _syncAttributes?: Attributes;
-  private _asyncAttributesPromise?: Promise<Attributes>;
-  private _attributes?: Attributes;
+  private _rawAttributes: [string, MaybePromise<AttributeValue | undefined>][];
+  private _asyncAttributesPending = false;
 
-  /**
-   * Check if async attributes have resolved. This is useful to avoid awaiting
-   * waitForAsyncAttributes (which will introduce asynchronous behavior) when not necessary.
-   *
-   * @returns true if the resource "attributes" property is not yet settled to its final value
-   */
-  public asyncAttributesPending?: boolean;
-
-  /**
-   * Returns an empty Resource
-   */
-  static empty(): IResource {
-    return Resource.EMPTY;
-  }
+  private _memoizedAttributes?: Attributes;
 
+  public static EMPTY = new Resource({});
```

Comment:
If in the future resources become mutable, would we be able to defend ourselves against this being mutated? 🤔

Reply:
I think the objects themselves won't be mutable. We'll have new resource objects emitted by a resource provider.